packages/tasit-action/src/TasitAct.js

```diff
@@ -0,0 +1,144 @@
+import "ethers/dist/shims.js";
+// Note: ethers SHOULD be imported from their main object
+// shims aren't injected with package import
+import { ethers } from "ethers";
+
+class Utils {
+  static isAddress = address => {
+    return typeof address === "string" && address.match(/^0x[0-9A-Fa-f]{40}$/);
+  };
+
+  static isABI = abi => {
+    return abi && Array.isArray(abi);
+  };
+
+  // https://github.com/ethers-io/ethers.js/blob/db383a3121bb8cf5c80c5488e853101d8c1df353/src.ts/utils/properties.ts#L20
+  static isEthersJsSigner = signer => {
+    return signer && signer._ethersType === "Signer";
+  };
+}
+
+class Subscription {
+  #txPromise;
+  #provider;
+
+  constructor(txPromise, provider) {
+    this.#txPromise = txPromise;
+    this.#provider = provider;
+  }
+
+  on = async (trigger, callback) => {
+    let blockHeight = 0;
+
+    if (trigger === "confirmation") blockHeight = 1;
+    else {
+      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+    }
+
+    if (!callback || typeof callback !== "function") {
+      throw new Error(`Cannot subscribe without a function`);
+    }
+
+    const tx = await this.#txPromise;
+
+    this.#provider.on(tx.hash, async receipt => {
+      const message = {
+        data: {
+          confirmations: receipt.confirmations,
+        },
+      };
+
+      try {
+        await callback(message);
+      } catch (error) {
+        throw new Error(`Callback function with error: ${error.message}`);
+      }
+    });
+  };
+
+  removeAllListeners = async () => {
+    const tx = await this.#txPromise;
```

Comment:
I think there should be a way to removeAllListeners without needing to await anything

Reply:
Yes, it is possible. I've changed the code to:

```javascript
if (this.#tx != undefined) this.#provider.removeAllListeners(this.#tx.hash);
```
When subscription is created, their constructor receive the promise of the sent tx. The promise will be solved inside of `on()` body and assigned to `tx`. If `tx` is `undefined` it's because no listening was made on this subscribe object.

Reply:
Nice! That sounds good.

Do we want to restrict it to there being only one tx at a time? For a given single subscription, I think that's the way to go, yes. I wonder if we want a convenience function for giving the same event handler function to all tx subscriptions.

Reply:
A) removeAllListeners (renamed to off) is no longer async
B) Extracted to https://github.com/tasitlabs/TasitSDK/issues/120

Reply:
Resolved this!